tracing-subscriber/src/registry/fmt_layer.rs

```diff
@@ -0,0 +1,197 @@
+    inner: PhantomData<S>,
+}
+
+impl FmtLayer {
+    pub fn builder() -> FmtLayerBuilder {
+        FmtLayerBuilder::default()
+    }
+}
+
+impl<S, N, E, W> FmtLayerBuilder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<N> + 'static,
+    W: MakeWriter + 'static,
+{
+    pub fn with_interest<F>(self, f: F) -> Self
+    where
+        F: Fn(&Event<'_>) -> bool + Send + Sync + 'static,
+    {
+        Self {
+            is_interested: Box::new(f),
+            ..self
+        }
+    }
+}
+
+// this needs to be a seperate impl block because we're re-assigning the the W2 (make_writer)
+// type paramater from the default.
+impl<S, N, E, W> FmtLayerBuilder<S, N, E, W> {
+    pub fn with_writer<W2>(self, make_writer: W2) -> FmtLayerBuilder<S, N, E, W2>
+    where
+        W2: MakeWriter + 'static,
+    {
+        FmtLayerBuilder {
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+            is_interested: self.is_interested,
+            inner: self.inner,
+            make_writer,
+        }
+    }
+}
+
+impl<S, N, E, W> FmtLayerBuilder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<N> + 'static,
+    W: MakeWriter + 'static,
+{
+    pub fn build(self) -> FmtLayer<S, N, E, W> {
+        FmtLayer {
+            is_interested: self.is_interested,
+            inner: self.inner,
+            make_writer: self.make_writer,
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+        }
+    }
+}
+
+impl Default for FmtLayerBuilder {
+    fn default() -> Self {
+        Self {
+            is_interested: Box::new(|_| true),
+            inner: PhantomData,
+            fmt_fields: format::DefaultFields::default(),
+            fmt_event: format::Format::default(),
+            make_writer: io::stdout,
+        }
+    }
+}
+
+fn name_of<T>(t: T) -> &'static str {
+    type_name::<T>()
+}
+
+// === impl Formatter ===
+
+impl<S, N, E, W> Layer<S> for FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn on_close(&self, id: Id, _: Context<'_, S>) {
+        // dbg!(id);
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        let mut writer = self.make_writer.make_writer();
+        write!(
+            &mut writer,
+            "{timestamp} {level} {target}",
+            timestamp = humantime::format_rfc3339_seconds(SystemTime::now()),
+            level = ColorLevel(event.metadata().level()),
+            target = &event.metadata().target(),
+        )
+        .unwrap();
+
+        let mut visitor = EventVisitor {
+            buf: String::new(),
```

Comment:
if we're going to write to a buffer instead of directly to a writer, we shouldn't allocate it every time we write an event. instead, we should reuse the thread-local buffer code from fmt that reuses a single allocation per thread.

Reply:
(i know this is a prototype, and the current code is fine _for now_, but i don't think it should be merged to master like this)

Reply:
agreed, this will be an issue. i'll reuse fmt's thread-local buffer.